File: src/test/java/ca/frar/jjjrmi/translator/GenerateJSON.java
package ca.frar.jjjrmi.translator;
import ca.frar.jjjrmi.exceptions.JJJRMIException;
import ca.frar.jjjrmi.translator.testclasses.*;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import org.json.JSONObject;

/**
 * Generate JSON encoding from the Java classes, placing the result in a file.
 * This get's called by the JS test
 * @author Ed Armstrong
 */
public class GenerateJSON {

    public static void main(String... args) throws IOException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        new GenerateJSON().run(args[0]);
    }

    JSONObject json = new JSONObject();

    private void run(String filename) throws IOException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        this.doGenerate();
        this.write(filename);
    }

    private void doGenerate() throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        Method[] declaredMethods = this.getClass().getDeclaredMethods();
        for (Method method : declaredMethods) {
            if (method.getName().startsWith("generate_")) {
                String key = method.getName().substring(9);
                TranslatorResult value = (TranslatorResult) method.invoke(this);
                if (value != null) this.json.put(key, value.toJSON());
            }
        }
    }

    private void write(String filename) throws IOException {
        File file = new File(filename);
        if (file.exists()) file.delete();
        file.getParentFile().mkdirs();
        file.createNewFile();

        FileOutputStream fos = new FileOutputStream(file);
        BufferedOutputStream bos = new BufferedOutputStream(fos);

        String jsonString = this.json.toString(2);
        bos.write(jsonString.getBytes());

        bos.close();
        fos.close();
    }

    public TranslatorResult generate_circular() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new CircularRef());
    }

    public TranslatorResult generate_referenceAsRoot() throws JJJRMIException {
        Translator translator = new Translator();
        Has<Object> has = new Has<>(null);

        this.json.put("hasRoot", translator.encode(has).toJSON());
        this.json.put("hasRef", translator.encode(has).toJSON());
        this.json.put("hasField", translator.encode(new Has<>(has)).toJSON());

        return null;
    }

    public TranslatorResult generate_handled() throws JJJRMIException {
        Translator translator = new Translator();
        HasHandler hasHandler = new HasHandler(2, 7);
        return translator.encode(hasHandler);
    }

    public TranslatorResult generate_nonEmptyArray() throws JJJRMIException {
        Translator translator = new Translator();
        int[] array = new int[]{1, 3, 7};
        return translator.encode(new Has<>(array));
    }

    public TranslatorResult generate_emptyArray() throws JJJRMIException {
        Translator translator = new Translator();
        int[] array = new int[0];
        return translator.encode(new Has<>(array));
    }

    public TranslatorResult generate_hasNull() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new Has<>(null));
    }

    public TranslatorResult generate_primitivesExtended() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new PrimitivesExtended(16));
    }

    public TranslatorResult generate_simple() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new Simple());
    }

    public TranslatorResult generate_none() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new None());
    }

    public TranslatorResult generate_primitives() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new Primitives(9));
    }

    public TranslatorResult generate_arrayWrapper() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new Primitives(9));
    }

    public TranslatorResult generate_doNotRetainExtends() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new DoNotRetainExtends(5));
    }

    public TranslatorResult generate_doNotRetainAnno() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new DoNotRetainAnno(5));
    }

    public TranslatorResult generate_transientField() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new TransientField().set(9));
    }

    public TranslatorResult generate_hasAfterDecode() throws JJJRMIException {
        Translator translator = new Translator();
        return translator.encode(new HasAfterDecode());
    }
}

File: src/test/js/GenerateJSON.js
/* global process, Reflect */
"use strict";
const fs = require("fs");
const Translator = require("../../main/js/translator/Translator");
const packageFile = require("./testclasses/packageFile");

function getAllMethodNames(obj) {
    let methods = new Set();
    while (obj) {
        let keys = Reflect.ownKeys(obj);
        keys.forEach((k) => methods.add(k));
        obj = Reflect.getPrototypeOf(obj);
    }
    return methods;
}

class GenerateJSON {
    constructor() {
        this.json = {};
    }
    run() {
        let methods = getAllMethodNames(this);
        for (let method of methods) {
            if (!method.startsWith("generate_")) continue;
            let r = this[method]();
            if (r) this.json[method.substring(9)] = r;
        }
    }
    build(method){
        let r = this[method]();
        if (r) this.json[method.substring(9)] = r;
    }

    generate_circular(){
        let translator = new Translator();
        return translator.encode(new packageFile.CircularRef());
    }

    generate_referenceAsRoot(){
        let translator = new Translator();
        let has = new packageFile.Has(null);

        this.json['hasRoot'] = translator.encode(has).toJSON();
        this.json['hasRef'] = translator.encode(has).toJSON();
        this.json['hasField'] = translator.encode(new packageFile.Has(has)).toJSON();

        return null;
    }

    generate_handled(){
        let translator = new Translator();
        translator.registerPackage(packageFile);
        return translator.encode(new packageFile.HasHandler(2, 7));
    }

    generate_nonEmptyArray(){
        let translator = new Translator();
        let array = [1, 3, 7];
        return translator.encode(new packageFile.Has(array));
    }

    generate_emptyArray(){
        let translator = new Translator();
        let array = [];
        return translator.encode(new packageFile.Has(array));
    }

    generate_hasNull(){
        let translator = new Translator();
        return translator.encode(new packageFile.Has(null));
    }

    generate_primitivesExtended(){
        let translator = new Translator();
        return translator.encode(new packageFile.PrimitivesExtended(16));
    }

    generate_simple(){
        let translator = new Translator();
        return translator.encode(new packageFile.Simple());
    }

    generate_none(){
        let translator = new Translator();
        return translator.encode(new packageFile.None());
    }

    generate_primitives(){
        let translator = new Translator();
        return translator.encode(new packageFile.Primitives(9));
    }

    generate_arrayWrapper(){
        let translator = new Translator();
        return translator.encode(new packageFile.Primitives(9));
    }
    
    generate_doNotRetainExtends(){
        let translator = new Translator();
        return translator.encode(new packageFile.DoNotRetainExtends(5));
    }
    
    generate_doNotRetainAnno(){
        let translator = new Translator();
        return translator.encode(new packageFile.DoNotRetainAnno(5));
    }        
    generate_transientField(){
        let translator = new Translator();
        return translator.encode(new packageFile.TransientField().set(9));
    }      
}

const filename = process.argv[2];
const method = process.argv[3];

if (filename.indexOf("/" !== -1)){
    let index = filename.lastIndexOf("/");
    let dir = filename.substr(0, index);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir);
}

let generateJSON = new GenerateJSON();

if (method) generateJSON.build(method);
else generateJSON.run();

console.log("writing JS data file to " + filename);
fs.writeFileSync(filename, JSON.stringify(generateJSON.json, null, 2));